src/ImageSharp/Common/Helpers/SimdUtils.HwIntrinsics.cs

```diff
@@ -246,6 +312,75 @@ public static class HwIntrinsics
                 }
             }
 
+            [MethodImpl(InliningOptions.ShortMethod)]
+            private static unsafe void Pad3Shuffle4(
+                ReadOnlySpan<byte> source,
+                Span<byte> dest,
+                byte control)
+            {
+                if (Ssse3.IsSupported)
+                {
+                    Vector128<byte> wMask = Vector128.Create(0xff000000u).AsByte();
+                    Vector128<byte> padMask = Vector128.Create(0, 1, 2, -1, 3, 4, 5, -1, 6, 7, 8, -1, 9, 10, 11, -1).AsByte();
+
+                    Span<byte> bytes = stackalloc byte[Vector128<byte>.Count];
+                    Shuffle.MmShuffleSpan(ref bytes, control);
+                    Vector128<byte> vcm = Unsafe.As<byte, Vector128<byte>>(ref MemoryMarshal.GetReference(bytes));
+
+                    fixed (byte* sBase = source)
+                    fixed (byte* dBase = dest)
+                    {
+                        byte* s = sBase;
+                        byte* d = dBase;
+
+                        // TODO: Consider unrolling and shuffling 4 at a time using Ssse3.AlignRight
+                        // See https://stackoverflow.com/questions/2973708/fast-24-bit-array-32-bit-array-conversion
+                        for (int i = 0; i < source.Length; i += 16)
+                        {
+                            Vector128<byte> vs0 = Sse2.LoadVector128(s);
```

Comment:
Same here, you'll over-read the input by 4 bytes at the end unless you stop short of the last vector -- or switch it up to handle 3 full input vectors at a time.

Reply:
And here's the other direction, simplified slightly from the C version in the SO answer:

```C#
unsafe static void widen24to32()
{
    var arr = new byte[48];
    for (int i = 0; i < arr.Length; i++)
        arr[i] = (byte)i;

    var vmask = Vector128.Create(0, 1, 2, 0x80, 3, 4, 5, 0x80, 6, 7, 8, 0x80, 9, 10, 11, 0x80).AsByte();
    var vfill = Vector128.Create(0xff000000ff000000ul).AsByte();

    fixed (byte* bp = &arr[0])
    {
        var v0 = Sse2.LoadVector128(bp);
        var v1 = Sse2.LoadVector128(bp + Vector128<byte>.Count);
        var v2 = Sse2.LoadVector128(bp + Vector128<byte>.Count * 2);
        var v3 = Sse2.ShiftRightLogical128BitLane(v2, 4);

        var v1a = Ssse3.AlignRight(v1, v0, 12);
        var v2a = Ssse3.AlignRight(v2, v1, 8);

        v0 = Sse2.Or(Ssse3.Shuffle(v0, vmask), vfill);
        v1 = Sse2.Or(Ssse3.Shuffle(v1a, vmask), vfill);
        v2 = Sse2.Or(Ssse3.Shuffle(v2a, vmask), vfill);
        v3 = Sse2.Or(Ssse3.Shuffle(v3, vmask), vfill);

        Console.WriteLine(v0);
        Console.WriteLine(v1);
        Console.WriteLine(v2);
        Console.WriteLine(v3);
    }
}
```

And the crappy diagram:

```
       v0               v1                  v2                 v3
BGRBGRBGRBGRBGRB GRBGRBGRBGRBGRBG    RBGRBGRBGRBGRBGR
                 palignr(v1, v0, 12) palignr(v2, v1, 8) psrldq(v2, 4)
                        ||                  ||                 ||
                        \/                  \/                 \/
BGRBGRBGRBGRxxxx BGRBGRBGRBGRxxxx    BGRBGRBGRBGRxxxx   BGRBGRBGRBGRxxxx
pshufb(mask)     pshufb(mask)        pshufb(mask)       pshufb(mask)
       ||               ||                  ||                 ||
       \/               \/                  \/                 \/
BGRxBGRxBGRxBGRx BGRxBGRxBGRxBGRx    BGRxBGRxBGRxBGRx   BGRxBGRxBGRxBGRx
```

Reply:
Ah, actually (_and I should have put a comment in the private method here_). I ensure this method is provided a maximum value from the source that is divisible by 16 then slice the source by 3/4 of that value to pass to the remaining non-simd version. 

I'll be giving your approach a try though as it should let me operates on more. Awesome!

https://github.com/SixLabors/ImageSharp/pull/1409/files#diff-0850002018264fb94c99126707650397c3e2c3c1b5d4cbe01860b363b9aa9912R104-R118

Reply:
Yep. Can't figure out the loop here.

Reply:
Ah, I read that backwards because the math is weird... it's actually overly conservative.  Let's say you've got a nice round 1200px RGB input (3600 bytes -- giving you 4800 bytes of RGBA output).  You start by making sure 3600 is divisible by 16, which it is.  Then you multiply that by 3/4 (don't use float math for that by the way) for the remainder slice, so you process the remainder starting at 2700 bytes, which is a full 300px short of the end.

Your SIMD loop is using the full 16-byte-rounded source length for the loop counter, incrementing that by 16, but then you increment your pointer by 12 each iteration.  That gives you 225 iterations in this case, so you only process 2700 bytes.  It matches with the remainder calculation from the call site, but it's well short of what it could do.

If you use the sample above, you can do away with the pointers and just use `ref Vector128<byte>` locals as you have in the other ones you've done.  You would read 3x `Vector128` and write 4x `Vector128`.  You'd want to make sure your input is a multiple of 48 bytes, but don't multiply that by 3/4 -- your output will just be 4/3 of your input, which you're already checking.

Reply:
Maaaaan do I feel sheepish. Maybe I should have given it another day after having a cold. Just implemented the two methods easily after taking nap. Both are 2x faster now 🚀

Reply:
That shuffling logic is always a brain twister...  I'm sure a cold doesn't help.  😄 
Glad you got it sorted!